### ISSUE TITLE
JVM hangs at the end with 100% cpu 

### ISSUE THREAD

**Issue by lbeuster** [aws/aws-iot-device-sdk-java-v2#139]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-iot-device-sdk-java-v2/issues) and didn't find any solution

**Known Issue**
- [x] I'm using ATS data type endpoint: the endpoint should look like `<prefix>-ats.iot.<region>.amazonaws.com`

**Describe the bug**
Since 1.2.14 the JVM hangs at the end of the process with one core with 100% cpu. Happens in Eclipse and Maven-CLI. No Java thread exists anymore (in the IDE) only the java-process itself hangs. I have to stop it in the task manager manually.
The problem is the aws-crt, because 1.2.15 works again if I use the aws-crt from 1.2.13.

**SDK version number**
1.2.14/15

**Platform/OS/Hardware/Device**
What are you running the sdk on?
OSX 11.2.2, Java8 and 11.

**To Reproduce (observed behavior)**
* build a mqtt client connection
* optional: connect or publish
* close
* finish JVM
-> JVM hangs with 100%

**Expected behavior**
JVM doesn't hang.

**Logs/output**
If applicable, add logs or error output.

To enable logging, set the following system properties:

*REMEMBER TO SANITIZE YOUR PERSONAL INFO*

```
-Daws.crt.debugnative=true
-Daws.crt.log.destination=File
-Daws.crt.log.level=Trace
-Daws.crt.log.filename=<path and filename>
```

**Additional context**
[aws.log](https://github.com/aws/aws-iot-device-sdk-java-v2/files/6076293/aws.log)

**Comment by bretambrose** [comment]
Can you attach a full code sample?  I am not seeing this issue using the included samples.

**Comment by lbeuster** [comment]
Here's a testcase: https://github.com/lbeuster/aws-iot-139.git
Broker url, cert and key are not included.

The 100% cpu only occurs if I create the ClientBootstrap with EventLoopGroup(1). If I use the default ClientBootstrap(null, null) the JVM still hangs at the end but without cpu.

**Comment by bretambrose** [comment]
I'm pretty sure I know what this is and we may want to approach a fix from multiple directions.

The root cause is that the SDK now waits in its jni exit function for all native threads to terminate, but in your sample, the event loop group native thread is still hanging around.  The why is a long and not-so-interesting story, but the Java CRT uses a full Java ref-count implementation to every native resource, which is different than what you're used to in Java (a single .close() destroys).  In your sample's case, that event loop group still has a positive reference count.

In the samples, you'll notice that we make extensive use of try-with-resource blocks.  The close invoked automatically there is actually critical both to resource management and proper ref-counting.  A resource can only "escape" the try-block it was created in if (1) you manually add a ref or (2) you make another resource which adds a ref (and it or something depending on it escapes), or (3) you don't use a try-block.

Is this a good programming experience or easy to work with? Nooooo it's not.  We've had a backlog item for a long time now to add some automatic janitor-esque garbage collection to all the CRT resources so that you no longer need to manage inscrutable ref counts, but it just hasn't been prioritized.

Because this is so hard to work with, I am currently of the opinion that (only in the case of the Java CRT) we should not block shutdown waiting for native threads.  It turns applications that don't manage the ref counts perfectly from working-but-shutdown-isn't-perfect (but it doesn't matter) to hanging-on-shutdown.  I will run this by the others and if there's no hard resistance, make a release with this behavior.

In the meantime, as a work around, you can try following the model of the samples with painstaking ref-count work via try-blocks and the like, but if you'd prefer to wait for a non-blocking shutdown, that's totally understandable.

**Comment by lbeuster** [comment]
Thanks for the analysis and you're totally right, I was completely confused about the close methods on the HostResolver and the builder and really didn't know if I should call them or not. ;)

I modified the above test and removed the useless mtls stuff and only used the default client bootstrap. The builder is also being closed. And I don't see any other references I create that I forget to close. But the JVM still hangs at the end. So for me it seems that I don't even have the possibility to enable a clean shutdown.

As a workaround working with the older version is also totally fine.

**Comment by bretambrose** [comment]
This issue should be fixed starting with 1.2.16
